src/discord_bot/llm.py

```diff
@@ -4,22 +4,76 @@
 by modifying the `base_url`.
 """
 
+import os
 import openai
+import time
+import fitz
+import gdown
 
 MAX_CHARACTERS = 2000
 QUESTION_CUT_OFF_LENGTH = 150
 RESERVED_SPACE = 50  # for other additional strings. E.g. number `(1/4)`, `Q: `, `A: `, etc.
 
 
+async def answer_question(model: str, question: str, server_url: str) -> list[str]:
+  """
+  Calls the LLM model with the specified question and server URL to get the answer.
+  """
+  try:
+    messages = await _call_llm(model, question, server_url)
+    return messages
+  except Exception as e:
+    return [str(e)]
+
+
 async def review_resume(model: str, url: str, server_url: str) -> list[str]:
-  # Download PDF, handle errors
-  # Parse PDF
-  # Send to LLM
-  # Return response
-  return ["NOT IMPLEMENTED YET"]
+  try:
+    output_path = download_pdf(url)
+    text = parse_pdf(output_path)
+    os.remove(output_path)
+
+    question = (
+      "You are a resume reviewer. Your tasks are:\n"
+      + "- Show sentences with incorrect grammars, and suggest a way to correct them.\n"
+      + "- Provide suggestions to improve the resume: \n\n"
+      + f"{text}"
+    )
 
+    messages = await _call_llm(model, question, server_url, attach_question_to_message=False)
+    return messages
+  except Exception as e:
+    return [str(e)]
 
-async def answer_question(model: str, question: str, server_url: str) -> list[str]:
+
+def download_pdf(url: str) -> str:
+  try:
+    if not os.path.exists("cache"):
+      os.mkdir("cache")
+
+    output_path = f"cache/{time.time()}.pdf"
+    gdown.download(url, output_path, fuzzy=True)
+
+    return output_path
+
+  except Exception as e:
+    raise RuntimeError(f"Error in downloading PDF: {e}")
+
+
+def parse_pdf(pdf_path: str) -> str:
+  try:
+    with fitz.open(pdf_path) as pdf:
+      text_list = []
+      for page in pdf:
+        text_list.append(page.get_text())
+      text = "\n\n".join(text_list)
+
+    return text
+
+  except Exception as e:
+    raise RuntimeError(f"Error in parsing PDF: {e}")
```

Comment:
pdf related utilities should be in a different module. This `llm.py` module should only contain LLM related things.